Docker deployment & integration tests  updates. (#125)
veraison/services: deployments/docker/README.md

build Veraison. It is used to compile the service binaries and their plugins
that are then used to construct the service containers. This image is also used
to spawn debug shells.

The `manager` image is used by the `veraison` frontend to, among other things,
access the volumes where the logs and provisioned values are stored.

## Configuration

The deployment may be configured by changing the settings inside
[deplyment.cfg](./deployment.cfg). See the comments inside that file for the
explanation of the individual configuration values.

### `make` Options

You can set the following variables to influence the behaviour while executing
make targets:

`DEBUG_HOST`: sets the host name for the debug container. This is useful if you
want to substitute the debug container in place of one of the service
containers.

`DOCKER_BUILD_FLAGS`: additional flags to be passed to Docker when building
various images. This is passed to all image build invocations, so should only
be used for globally-applicable flags such as `--no-cache`.

## Debugging

Running `make debug` will pop up a "debug" shell inside the builder image. The
image has the build toolchain and the debugger installed, and

> `/veraison/build/` directory s mapped onto the root of Veraison repo on your

Reviewer: ```suggestion
`/veraison/build/` directory is mapped onto the root of Veraison repo on your
```

host. This means  you can make modifications to the source using your preferred
editor/IDE on the host, and then build and debug them in the debug shell.

### Example: debugging scheme verification plugin

For example, say you are implementing a new attestation scheme, and are running
into issues when attempting to verify your token (e.g. via an integration
tests test-case). There is not enough output in the logs to identify the issue,
and so, you'd like to set a break point inside your code.

> This  needs to be done inside `vts-serivice`, as this is the component that

Reviewer: typo: vts-service

loads the evidence handler plugin used on the verification path.

First, assuming you have frontend set up, and the services running, (if not,
you can do so with `source env.bash; veraison start`), you will need to stop
the "production" VTS service with:

    veraison stop vts

this will stop `vts-service` but should leave the other services running (you
can verify that by running `veraison status`).

Next, enter the debug shell:

    make DEBUG_HOST=vts-service debug

this will pop open a `bash` shell inside the builder. The `DEBUG_HOST` argument
will set the hostname of the builder container. Here, we're setting it to the
hostname of the VTS service we just stopped, so that the provisioning and
verification services that are still running will now talk our debug container
instead.

Next, navigate to the location of the VTS service executable (keeping in mind
that the root of the repo is mapped to `/veraison/build` inside the container),
and start the debugger:

    # inside the debug shell:
    cd /veraison/build/vts/cmd/vts-service
    debug

> 

Reviewer: here I'm getting this from delve:
```
$ debug
error obtaining VCS status: exit status 128
	Use -buildvcs=false to disable VCS stamping.
exit status 1
```
It'd look benign but I have never seen that, have you bumped into anything similar before?

Use the `-h` option to see the full list of available commands.

## Deployment Composition

The deployment exists on its own Docker network. Each of the three services
runs in its own container. Service logs and VTS stores reside inside docker
volumes, independent of any container.

![deployment composition](./misc/veraison-docker-deployment.png)

On the diagram above, the network is shown in orange, the volumes in purple,
the containers in blue, and the images the containers are based on in green.

In addition to the service container images, there also two images without
permanent containers. They are used to spawn ephemeral containers via `docker
run --rm`.

The `builder` image contains the Go toolchain and other tools necessary to
build Veraison. It is used to compile the service binaries and their plugins
that are then used to construct the service containers. This image is also used
to spawn debug shells.

The `manager` image is used by the `veraison` frontend to, among other things,
access the volumes where the logs and provisioned values are stored.

## Configuration

The deployment may be configured by changing the settings inside

> [deplyment.cfg](./deployment.cfg). See the comments inside that file for the

Reviewer 2: ```suggestion
[deployment.cfg](./deployment.cfg). See the comments inside that file for the
```

explanation of the individual configuration values.

### `make` Options

You can set the following variables to influence the behaviour while executing
make targets:

`DEBUG_HOST`: sets the host name for the debug container. This is useful if you
want to substitute the debug container in place of one of the service
containers.

`DOCKER_BUILD_FLAGS`: additional flags to be passed to Docker when building
various images. This is passed to all image build invocations, so should only
be used for globally-applicable flags such as `--no-cache`.

## Debugging

Running `make debug` will pop up a "debug" shell inside the builder image. The
image has the build toolchain and the debugger installed, and
`/veraison/build/` directory is mapped onto the root of Veraison repo on your

> host. This means  you can make modifications to the source using your preferred

Reviewer 2: ```suggestion
host. This means you can make modifications to the source using your preferred
```

editor/IDE on the host, and then build and debug them in the debug shell.

### Example: debugging scheme verification plugin

For example, say you are implementing a new attestation scheme, and are running
into issues when attempting to verify your token (e.g. via an integration
tests test-case). There is not enough output in the logs to identify the issue,

> and so, you'd like to set a break point inside your code.

Reviewer 2: ```suggestion
and so, you'd like to set a breakpoint inside your code.
```

This needs to be done inside `vts-service`, as this is the component that
loads the evidence handler plugin used on the verification path.

First, assuming you have frontend set up, and the services running, (if not,
you can do so with `source env.bash; veraison start`), you will need to stop
the "production" VTS service with:

    veraison stop vts

this will stop `vts-service` but should leave the other services running (you
can verify that by running `veraison status`).

Next, enter the debug shell:

    make DEBUG_HOST=vts-service debug

this will pop open a `bash` shell inside the builder. The `DEBUG_HOST` argument
will set the hostname of the builder container. Here, we're setting it to the
hostname of the VTS service we just stopped, so that the provisioning and

> verification services that are still running will now talk our debug container

Reviewer 2: ```suggestion
verification services that are still running will now talk to our debug container
```

instead.

Next, navigate to the location of the VTS service executable (keeping in mind
that the root of the repo is mapped to `/veraison/build` inside the container),
and start the debugger:

    # inside the debug shell:
    cd /veraison/build/vts/cmd/vts-service
    debug

The `debug` command is an alias for `dlv debug` that will make sure that the
debug executable will be built with evidence handling plugins compiled in (same
as if you ran `make SCHEME_LOADER=builtin`). This means you can set the

> breakpoint inside your code directly form here, and you don't need to worry

Reviewer 2: ```suggestion
breakpoint inside your code directly from here, and you don't need to worry
```
